### ISSUE TITLE
Referer: None for external links

### ISSUE THREAD

**Issue by ProgerXP** [dillbyrne/random-agent-spoofer#110]
Currently we have the choice of sending Referer either by default (as usual) or not sending at all. However, IMO the most useful scenario is not to send Referer to external resources because sometimes it is required when browsing inside the same resource. Besides, it doesn't look logical to limit Referer for the site itself because you are still browsing it and this is reflected in the logs. Your path is still easily mapped because of pages you have requested. Not so with external resources which you may follow from a link on some other resource.

To sum up, is it possible to send Referer as usual except when a cross-domain request occurs, for which the Referer is not sent (like *None* option)? Cross-domain means both image/CORS/other requests as well as regular page opening from a link.

**Comment by Mosrite** [comment]
Maybe you could add the same options as "Smart Referer" (https://addons.mozilla.org/en-US/firefox/addon/smart-referer/?src=search)? (Except for user given referer maybe and whitelist source).
I think the best setting would be to send the URL you're going to as referer, but not to treat subdomains as different domains. This works most of the time and prevents tracking.
(And we could get rid of an additional addon... ;) )

**Comment by ProgerXP** [comment]
> I think the best setting would be to send the URL you're going to as referer, but not to treat subdomains as different domains. 

That's great idea - to send Referer to resources within the current high-level domain (2nd level) but send blank Referer to all others. Simple to implement too.

**Comment by dillbyrne** [comment]
@Mosrite @ProgerXP Sure that could work, care has to be taken that the referer method used does not make the user stand out. As I said I will look into it and I will update this issue when I do.

**Comment by Mosrite** [comment]
Apparently, all important referer-related settings can be toggled in the about:config section:
http://www.ghacks.net/2015/01/22/improve-online-privacy-by-controlling-referrer-information/
Maybe you could add all these options (or at least the ones that make sense to modify) to the UI? At least, the "network.http.referer.XOriginPolicy" option should be added to the current options, I think. I don't know how difficult it is to cover these options with whitelist settings, too.
There is a whitelist by the author of the "Smart Referer" addon: http://meh.github.io/smart-referer/whitelist.txt Maybe some of this could be implemented in the default whitelist?

**Comment by alct** [comment]
For the record (could be useful resource) https://addons.mozilla.org/en-us/firefox/addon/smart-referer/

**Comment by dillbyrne** [comment]
@alct I can easily add options for the built in headers . I'll do it  after I get 0.9.5.4 signed by mozilla and released.

**Comment by Mosrite** [comment]
Will this come with an option to whitelist certain referers?

**Comment by dillbyrne** [comment]
@Mosrite I had planned to remove my implementation of referer blocking in place of the prefs mentioned above in which case there would be no option to whitelist it but I'll see what it behaves like when I implement it and get back to you. I can most likely keep the whitelisting.

**Comment by Mosrite** [comment]
I thought maybe you could just import the functions of the smart referer addon? I have no idea about coding though, so maybe that's a dumb idea... ;)

**Comment by dillbyrne** [comment]
I'm working on this at the minute . I have decided to make use of the mozilla referer prefs for spoofing source , x origin policy and trimming policy.  This will allow users to set it how they wish. I'm keeping the option to block or deny all also. In the event a referer is blocked but whitelisted the other preferences will take over. 

the new options are 

`network.http.referer.XOriginPolicy`: 0=always send, 1=send if base domains match, 2=send if hosts match
`network.http.referer.spoofSource`: false=real referer, true=spoof referer (use target URI as referer)
`network.http.referer.trimmingPolicy`: 0=send full URI, 1=scheme+host+port+path, 2=scheme+host+port


![r1](https://cloud.githubusercontent.com/assets/2903711/8429300/3d3eff60-1f1e-11e5-9c17-773eb791a0d5.png)

![r2](https://cloud.githubusercontent.com/assets/2903711/8429302/3d418f64-1f1e-11e5-9524-08339ed6bf71.png)

![r3](https://cloud.githubusercontent.com/assets/2903711/8429301/3d40a554-1f1e-11e5-966e-a7ad2332c0de.png)


I'll update when I have it done

**Comment by dillbyrne** [comment]
This will be available in the 0.9.5.4 release
